Fixed duplicate route creation issue in QuickStartCommand

erhanurgun opened the pull request:
Hello,

I've identified an issue in the QuickStartCommand and created this pull request to address it. Here are the changes I've made:

1. In `QuickStartCommand.php`:
   - Added a new method called `updateApiRoutes()`. This method checks if the 'users' route already exists and only adds it if it doesn't.
   - This new method inspects the routes/api.php file and only appends the new route when necessary.

2. Added `QuickStartCommandTest.php`:
   - Included a test method to ensure the command creates the correct files.
   - Added a test method to verify that API routes are updated correctly.
   - Included a test method to check that routes are not duplicated when the command is run multiple times.

These changes will ensure the command operates more reliably and reduce the likelihood of similar issues occurring in the future.

I'm looking forward to your feedback and review. If you have any questions or suggestions, please don't hesitate to ask.

Thank you!

GautierDele commented:
Hello @erhanurgun,

Thanks for your proposal, seems like a good idea
It seems that your tests arnt passing, could you have a look at it please ?

erhanurgun commented:
Hello @GautierDele,

Thank you for your feedback. I've reviewed the tests and made some adjustments to address the failing tests. Here's a summary of the changes I've made:

1. In `./phpunit.xml`:
   - Added environment variables for testing:
     ```xml
     <env name="APP_ENV" value="testing"/>
     <env name="DB_CONNECTION" value="testing"/>
     ```

2. In `./src/Console/Commands/QuickStartCommand.php`:
   - Updated the file creation logic:
     ```php
     file_put_contents($routesPath, '<?php');
     ```

3. In both `./tests/Features/Commands/QuickStartCommandTest.php` and `./tests/Feature/TestCase.php`:
   - Added a check to ensure `api.php` exists before running tests:
     ```php
     if (! File::exists(base_path('routes/api.php'))) {
         File::put(base_path('routes/api.php'), '<?php');
     }
     ```

4. In `./tests/Feature/TestCase.php`:
   - Added `getPackageProviders` and `getEnvironmentSetUp` methods to properly configure the testing environment.

These changes aim to resolve the "File does not exist at path .../routes/api.php" error that was causing the tests to fail.

I've pushed these updates to my fork. Could you please pull the changes and run the tests again? If you encounter any issues or have any questions, please let me know, and I'll be happy to assist further.

Thank you for your patience and collaboration.

Best regards,
<br>

![fixed_ss](https://github.com/Lomkit/laravel-rest-api/assets/20555797/77c6bdf8-087d-40cf-86f9-9d428c9558f1)

GautierDele commented on tests/Feature/Commands/QuickStartCommandTest.php:
```diff
@@ -0,0 +1,91 @@
+        $this->assertStringContainsString('\App\Models\User::class', $resourceContent);
+
+        // Check for the updated namespace in UsersController.php
+        // TODO: We are putting this control in the comment line for now, because UsersController may not have this change made.
+        // $this->assertStringContainsString('\App\Models\User::class', $controllerContent);
```
The controller needs to have the resource, not the model ? Might be why it's not working

GautierDele commented on tests/Feature/TestCase.php:
```diff
@@ -16,11 +16,55 @@ protected function setUp(): void
     {
         parent::setUp();
 
+        $this->loadMigrationsFrom(__DIR__.'/Support/Database/migrations');
+        $this->artisan('migrate', ['--database' => 'testing'])->run();
```
I'm confused about this because migrations were going great on my side, you might not have PHP UNIT configured nicely in your local environment

GautierDele commented on phpunit.xml:
```diff
@@ -18,6 +18,8 @@
     <env name="QUEUE_CONNECTION" value="sync"/>
     <env name="SESSION_DRIVER" value="array"/>
     <env name="TELESCOPE_ENABLED" value="false"/>
+    <env name="APP_ENV" value="testing"/>
+    <env name="DB_CONNECTION" value="testing"/>
```
Don't think we need this, so anyone can create his own testing settings

GautierDele commented on tests/Feature/TestCase.php:
```diff
@@ -16,11 +16,55 @@ protected function setUp(): void
+        ];
+    }
+
+    protected function getEnvironmentSetUp($app)
+    {
+        $app['config']->set('database.default', 'testing');
+
+        $connection = env('DB_CONNECTION', 'sqlite');
+
+        switch ($connection) {
+            case 'mysql':
+                $app['config']->set('database.connections.testing', [
+                    'driver' => 'mysql',
+                    'host' => env('DB_HOST', '127.0.0.1'),
+                    'port' => env('DB_PORT', '3306'),
+                    'database' => 'rest',
+                    'username' => env('DB_USERNAME', 'root'),
+                    'password' => env('DB_PASSWORD', ''),
+                ]);
+                break;
+            case 'pgsql':
+                $app['config']->set('database.connections.testing', [
+                    'driver' => 'pgsql',
+                    'host' => env('DB_HOST', '127.0.0.1'),
+                    'port' => env('DB_PORT', '5432'),
+                    'database' => 'rest',
+                    'username' => env('DB_USERNAME', 'postgres'),
+                    'password' => env('DB_PASSWORD', 'postgres'),
+                ]);
+                break;
+            default:
+                $app['config']->set('database.connections.testing', [
+                    'driver' => 'sqlite',
+                    'database' => ':memory:',
+                ]);
+        }
+    }
```
Might not be needed also, you can configure PHP UNIT on runtime for github actions and change the settings in your case

GautierDele commented on tests/Feature/Commands/QuickStartCommandTest.php:
```diff
@@ -0,0 +1,91 @@
+        $this->artisan('rest:quick-start')->assertExitCode(0);
+
+        // Check if the files were created
+        $this->assertFileExists(app_path('Rest/Resources/UserResource.php'));
+        $this->assertFileExists(app_path('Rest/Controllers/UsersController.php'));
```
I'm wondering what this is doing in test, i'm gonna have a look 😄

GautierDele commented:
I'm gonna have a quick look later this afternoon in order to test it on my side, thanks for your time
I'm gonna also fix the problem i gave above, so you don't need to 😉

GautierDele commented:
Thanks @erhanurgun merging !
